### PR TITLE
UAR-794: Add trusts section to review update statement page

### PR DESCRIPTION
**Pull request by conuirchgov**
### JIRA link

https://companieshouse.atlassian.net/browse/UAR-794

### Change description

Added trusts summary section to Review Update Statement page (behind Update Trusts feature flag)

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Review by joshuahigginsonCH — APPROVED** [review]

**Review by cdrainCH — APPROVED** [review]

File: src/utils/check.your.answers.ts [hunk 1]
```diff
@@ -88,3 +100,21 @@ export const postDataForReview = async (req: Request, res: Response, next: NextF
...
+const getBackLinkUrl = (isNoChangeJourney: boolean, hasAnyBosWithTrusteeNocs: boolean) => {
+  if (isNoChangeJourney) {
+    return UPDATE_DO_YOU_WANT_TO_MAKE_OE_CHANGE_URL;
+  } else {
+    const updateTrustsEnabled = isActiveFeature(FEATURE_FLAG_ENABLE_UPDATE_TRUSTS);
+
+    return updateTrustsEnabled && hasAnyBosWithTrusteeNocs
+      ? UPDATE_TRUSTS_ASSOCIATED_WITH_THE_OVERSEAS_ENTITY_URL
+      : UPDATE_BENEFICIAL_OWNER_TYPE_URL;
+  }
```
Comment (AndreKramer7654) [inline src/utils/check.your.answers.ts:113]: what about registration?

File: src/utils/update/trusts.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { ApplicationData } from "model";
+
+export const checkHasAnyBosWithTrusteeNocs = (appData: ApplicationData) => {
+  const bosThatCanHaveTrusteeNoc = [
+    ...(appData.beneficial_owners_individual || []),
+    ...(appData.beneficial_owners_corporate || []),
+    ...(appData.update?.review_beneficial_owners_individual || []),
```
Comment (AndreKramer7654) [inline src/utils/update/trusts.ts:7]: we should not be checking the review ones. They are temporary and moved into main list once reviewed!

File: src/utils/update/trusts.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { ApplicationData } from "model";
+
+export const checkHasAnyBosWithTrusteeNocs = (appData: ApplicationData) => {
+  const bosThatCanHaveTrusteeNoc = [
```
Comment (AndreKramer7654) [inline src/utils/update/trusts.ts:4]: quite a bit of work to join the lists into one big one just to check some. Could just use 
[appData.beneficial_owners_individual, appData.beneficial_owners_corporate].some(l => l.map(bo => bo.trustees_nature_of_control_types > 0));

File: src/utils/check.your.answers.ts (hunk 1 above)
Comment (AndreKramer7654) [reply src/utils/check.your.answers.ts:113]: ok - not used in registration

File: src/utils/update/trusts.ts (hunk 2 above)
Comment (conuirchgov) [reply src/utils/update/trusts.ts:7]: Inclined to leave as is - does no harm in checking both.

And in the no change journey, where this is currently called, the beneficial owners are actually still in the review lists.

File: src/utils/update/trusts.ts (hunk 3 above)
Comment (conuirchgov) [reply src/utils/update/trusts.ts:4]: I don't see the benefit of flatMap here either. Can you write up a suggestion? 

Tbh I think `[...(appData.array || []), ...]` is the simplest way of flattening those arrays.

File: src/utils/update/trusts.ts (hunk 2 above)
Comment (AndreKramer7654) [reply src/utils/update/trusts.ts:7]: ok, it's important to know. We should really have separate checks for each set of lists. review and reviewed

File: src/utils/update/trusts.ts (hunk 3 above)
Comment (AndreKramer7654) [reply src/utils/update/trusts.ts:4]: [ appData.array1, appData.array2].flat().some(a => a?.length > 0);

was hoping this handles undefines so no need for "|| []"

Comment (AndreKramer7654) [reply src/utils/update/trusts.ts:4]: typescript flat() does not seem to deal well with undefines. Some languages just filter those. So how about the some of some suggestion. Just to iterate over the list - not expand everything into a new list as [ ...array1, ...array2 ] does:

type X = {
   t? : Array<String>;
};

const l0 : Array<X> = [];
const l1 : Array<X> = [ {} ];
const l2 : Array<X> | undefined = undefined;
const l3 : Array<X> = [{ t: ["here "] }];

const result = [l0, l1, l2, l3].some(x => (x ?? []).some(y => ( y.t ?? []).length > 0));

console.log(result);

File: src/utils/update/trusts.ts (hunk 2 above)
Comment (AndreKramer7654) [reply src/utils/update/trusts.ts:7]: Maybe just go over review lists then?

File: src/utils/update/trusts.ts (hunk 3 above)
Comment (conuirchgov) [reply src/utils/update/trusts.ts:4]: I don't think this solution is as readable, and if there's an implied argument for potential performance improvements, I'd expect the benefit to be negligible.

Gonna make a call here and say I don't think this change would provide any real benefit.

File: src/utils/update/trusts.ts (hunk 2 above)
Comment (conuirchgov) [reply src/utils/update/trusts.ts:7]: > ok, it's important to know. We should really have separate checks for each set of lists. review and reviewed

Should we? Currently no need for checking them independently here. 

> Maybe just go over review lists then?

As above, I'm inclined to leave it as is. Right now we just need to know if there are _any_ BOs with trustee NOC; where they are doesn't matter.

**Review by stephanie-ch — APPROVED** [review]

File: src/utils/update/trusts.ts (hunk 3 above)
Comment (AndreKramer7654) [reply src/utils/update/trusts.ts:4]: There is duplication with scr/utils/trust.ts which checks for NoC too.

**Pull request merged**